### DISCUSSION TITLE
Way to improve RefinedTypeOps?

### DISCUSSION THREAD

**Discussion by rlemaitre** [Iltotore/iron#226] [Ideas]
I encountered [this](https://contributors.scala-lang.org/t/scala-3-macro-annotations-and-code-generation/6035/69) and wonder if the same strategy could be used to improve `RefinedTypeOps` declaration.

WDYT?

**Comment by Iltotore** [comment]
I need to investigate the solution further but it indeed looks promising. Looks like it is possible by using making `A`/`C`/`T` type parameters actual generated type members but some parts remain unclear:
- How would this play with the `inline` machinery that powers Iron?
- As said in the post you linked, it does not (yet) play well with Intellij IDEA
- Is it possible to get the underlying type of an opaque alias? This might be the main blocker
- How to introduce it in a retrocompatible way (or without breaking retrocompatibility too much)?

And last one: do we actually need such solution? We don't need to generate new definitions: we only need to derive implementations. For instance, `RefinedTypeOps` could look like this:
```scala
trait RefinedTypeOps[T]: //T is the opaque (or transparent) alias

  type BaseType = deriveBaseType[T]

  type ConstraintType = deriveConstraintType[T]
```

but here comes another problem: macros (AFAIK) cannot return a type.

**Reply by rlemaitre** [reply]
You're right, maybe this isn't something we need for `RefinedTypeOps` but it can be useful some day for another thing 😄
